Proposal for Documentation Restructuring (#77)
prometheus-operator/website: content/docs/proposal/202405-docs-enhancement.md

# Revamping Documentation Structure

- Owners:
- [AshwinSriram11](https://github.com/AshwinSriram11)
- Related Tickets:
- Other docs:
- N/A

This document aims to restructure the documentation of Prometheus Operator website for improving content organization.

# Why

It is important to keep the documentation of a project updated with the development process so that all features, enhancements are incorporated in the documentation properly. For that, the basic structure for the documentation should be intuitive so that further additions to the documentation is easier. 
# Pitfalls of the current solution

* Disorganized arrangement of topics due to which it becomes difficult to search for essential information according to the requirement.
* Navigating through the current structure becomes tedious and decreases productivity.

> * Lack of clarity according to the various user personas, such as users, administrators and contributors.

Author: I made a little change but it is still not finished.

# Goals

* To streamline navigation process by intuitive structuring.

> * Proper segregation of information according to the persona.

Reviewer: Adjusting information by personas doesn't sound like our goal, it sounds more like how we will reach our goal.

As someone inexperienced with Prometheus-Operator, what do you think when opening the website? Is it "`Oh, I wish this website was organized by the different personas that will interact with the operator!`"? Or is it "`I wish I could find the information I want ASAP so I can go back to what I was doing earlier`"?

It might be ok to briefly mention how we want to reach our goals in this section, but the focus is the problems we want to solve! How we'll solve them will be detailed in the `How` section

Reviewer: I also noticed how you don't talk about the development process in our goals, but you mention the problem in the Why section. I'm a bit confused 😬

# Audience

* Developers/Engineers who rely on Prometheus Operator to facilitate the deployment, management and monitoring of Prometheus instances in their environment.
* Administrators who are responsible for optimizing Prometheus Operator.

> * Contributors who want to engage and want to contribute effectively to the project.

Reviewer: Looks great!

What about Alertmanager administrators? And application developers who want to set up alerts about their apps?

# Non-Goals

> * This proposal only aims for improving content organization of the website.

Reviewer: The first line is talking about one of our goals, it doesn't seem to fit in "Non-Goals"

> * It does not address other documentation issues regarding adding information about new feature, enhancement, troubleshooting, etc. 

Reviewer: ```suggestion
* Adding extra documentation about features, enhancements, troubleshooting, or other information that might be missing today. 
```

# How

I believe a basic structure for the website should be as given below:
* Getting Started
  * Introduction
  * Installation Guide
* API Reference
* User Guide
* Administrator Guide
* Community
  * Contributing
  * Testing
  * DCO
  * License
  * Code of Conduct
  * Release

>   * Adopters

Reviewer: Those sections look great!

Could you provide a bit more context about it? Maybe replace the bulleted list with more details on what to expect in each section?

I also think that Kube-Prometheus should be removed from the docs because:
* It is a seperate project from Prometheus Operator.

> * It is not currently being supported by the community so it gives rise to more issues.

Reviewer: Actually, Kube-Prometheus is a project that lives in the prometheus-operator organization (https://github.com/prometheus-operator/kube-prometheus), and is maintained by the @prometheus-operator/kube-prometheus-reviewers. 

The confusion is that there is yet another project called [kube-prometheus-stack](https://github.com/prometheus-community/helm-charts/tree/main/charts/kube-prometheus-stack), which indeed is a separate project but is arguably more popular than the one that lives in our org.

With that said, both projects are popular ways of installing a full monitoring stack alongside prometheus-operator. Does it make sense to include them in the "Installing" section?

Author: My bad! I didn't check about this properly.

I think installation should be included in the Installation-Guide page. Other than that, where should the Kube-Prometheus section be placed at? I am confused about this.

For now, will remove what I wrote about Kube-Prometheus.

Reviewer: Sorry, I didn't mean to say that we **_should_** have a kube-prometheus section. I'm ok with removing the kube-prometheus section and just mentioning it as one of the installation options.

What I meant is that, if we're going to remove it, let's write the correct information in the proposal and clear explanation on why we think it's a good idea :)

Author: No need to apologise . So I think we should mention that helm charts are not being used nor maintained by Prometheus-Operator and when we add the installation method, we will also need to provide a warning.

What do you think?

Reviewer: No need to apologize! Personally, I think it's worth mentioning the helm charts because we can't ignore their popularity. Mention that it's maintained by a separate set of folks is also a good a idea

ccd# Revamping Documentation Structure

- Owners:
  - [AshwinSriram11](https://github.com/AshwinSriram11)
- Related Tickets:
- Other docs:
  - N/A

This document proposes restructuring the documentation of the Prometheus-Operator website, aiming for better content organization and better experience for new users and old veterans.

# Why

Restructuring the documentation will help in improving user experience and may save time for a user to search for relevant information effectively. This will encourage a newcomer to get familiar with the project in an efficient manner. The end goal of a good documentation is to follow the industrial best practices and provide accurate information to a user. If the basic structure of a documentation is improved, then further improvements in the future would become easier.

# Pitfalls of the current solution

> A good documentation is one that is easy to understand for a newcomer and provides exact amount of information that is needed according to the need. But looking at the current documentation structure, a lot of topics seem misplaced. For example, there is no need of "Contributing" page in the prologue section. Prologue section should only give the introduction and the pre-requisites for the project. Due to this, a user might need to put more effort to search for relevant information and decrease user's productivity.

Reviewer 2: I think is a good idea to add few more examples on the problems of the existing solution and maybe even link to existing issues of missing contents for example.

wdyt?

Author: Ya sounds like a good idea.

Flow of content should also be uniform. 

* Disorganized arrangement of topics due to which it becomes difficult to search for essential information according to the requirement.
* Navigating through the current structure becomes tedious and decreases productivity.
* Lack of clarity according to the various user personas, such as users, administrators and contributors.

# Goals

* To improve readability and make the documentation easier to understand.

> * To save user's time by streamlining navigation process by intuitive structuring.

Reviewer 2: I think it's explicity by the first point, I'm a bit worried about mentioning measurements such as time improvements since it's very hard to measure.

Because if we mention our goal is to save time we first need to understand what's the spending time from users reading the docs and later understand how we can compare our implementation in terms of time saving.

I'd prefer to have goals which are less subjectives, for example ensure we have all the features of the latest version documented (just an example)

Its easy to measure, we can raise all the features from the operator and later ensure we have a doc for each one.

* Developers who want to setup alerts for their application.
* Administrators who want to monitor Prometheus and Alertmanager instances using Kubernetes tools.
* Maintainers who are responsible for optimizing Prometheus Operator.
* Contributors who want to engage and want to contribute effectively to the project.

# Non-Goals

* Adding extra documentation about new features, enhancements, troubleshooting, or other information that might be missing today. 

# How

I believe a basic structure for the website should be as given below:
### 1. Getting Started
  * #### Introduction
     This section will introduce us to Prometheus-Operator and will talk about goals of this project and the problems that are being solved by this project.
  * #### Installation Guide
     This section will contain all the methods of installation of Prometheus-Operator in you local environment.
### 2. API Reference
   This section will provide detailed information about different fields of the Custom Resources in Prometheus-Operator(config, spec, status and other information).    
### 3. User Guide

>    This section will provide focused information for developers who want to integrate Prometheus and Alertmanager using Prometheus-Operator in their applications along with examples and trobleshooting.

Reviewer 2: I was a bit confused about this section, because of the following text: `developers who want to integrate Prometheus and Alertmanager using Prometheus-Operator in their applications`

Are we going to talk about instrumentation on this section?

Reviewer: I don't think so. @AshwinSriram11 can confirm, but I think he was mentioning the usage of Service/PodMonitor to make sure Prometheus understands where their metrics are. I can also see app developers being responsible for setting up their PrometheusRules, so that's beyond the scope of Prometheus Administrators

I believe that by restructuring, we can group the related topics together in an organized way and use cross-references to link topics that are difficult to group but are relevant for understanding. This makes the documentation more user-friendly and also ensures uniform flow of content.

After we have a proper structure, it will become relatively easy to add information about new features. Maintainers will be able to save time on deciding the best place for adding new content. Also, improving documentation will help in improving Search Engine Optimization(SEO) and increase the engagement of the project.

# Pitfalls of the current solution

A good documentation is one that is easy to understand for a newcomer and provides the exact amount of information that is needed according to the need. But looking at the current documentation structure, a lot of topics seem misplaced. For example, there is no need of **"Contributing"** page in the prologue section. Prologue section should only give the introduction and the prerequisites for the project. Due to this, a user might need to put more effort to search for relevant information and decrease the user's productivity.

> Currently, there is some unnecessary information from the documentation creating misconceptions in the user's mind. For example, let us look at [#6046](https://github.com/prometheus-operator/prometheus-operator/issues/6046) which tells us that there is no Ingress Guide present in the current documentation. But, if we look at the website, there are links to the **“Ingress Guide”** on the [Getting-Started](https://prometheus-operator.dev/docs/user-guides/getting-started/#exposing-the-prometheus-service) and [Alerting page](https://prometheus-operator.dev/docs/user-guides/alerting/#exposing-the-alertmanager-service) in **User-Guide**. Due to this, many users will report the same issue and it will take time for maintainers to resolve them.

Reviewer: Wow, I didn't know we had links to unexisting pages on our website :O

# Goals

* To improve readability and make the documentation easier to understand.
* To ensure a uniform flow of content. 
* To remove irrelevant information from the documentation.
* To ensure that all new features get documented properly in the future.

# Audience

* Developers who rely on Prometheus Operator to facilitate the deployment, management and monitoring of Prometheus instances in their environment.
* Developers who want to setup alerts for their application.
* Administrators who want to monitor Prometheus and Alertmanager instances using Kubernetes tools.
* Maintainers who are responsible for optimizing Prometheus Operator.
* Contributors who want to engage and want to contribute effectively to the project.

# Non-Goals

> * Adding extra documentation about new features, enhancements, troubleshooting, or other information that might be missing today. 

Reviewer: Maybe we could mention this as a follow up?  I.e., if we finish the project and there's still time to do more, we could start documenting missing features

# How

I believe a basic structure for the website should be as given below:
### 1. Getting Started
  * #### Introduction
     This section will introduce us to Prometheus-Operator and will talk about goals of this project and the problems that are being solved by this project.
  * #### Installation Guide

>      This section will contain all the methods of installation of Prometheus-Operator in local environment.

Reviewer 2: Agree with @ArthurSens, besides that the instalation guide is K8S Like so it doesn't really matter if it's a kind or a EKS.

### 2. API Reference

>    This section will provide detailed information about different fields of the Custom Resources in Prometheus-Operator(config, spec, status and other information).    

Reviewer: The API reference is being moved from a sub section to its dedicated section. Could you explain your thoughts on why we're making this change?

Author: I have added explanation for each section below.

### 3. User Guide
   This section will provide focused information for developers who want to integrate Prometheus and Alertmanager using Prometheus-Operator in their applications along with examples and trobleshooting.
### 4. Administrator Guide

>    This section will provide deeper insights for administrators who want to setup a robust, secure monitoring infrastructure. 

Reviewer: Thinking of a "uniform flow of content", do you think people will read first the "User guide" and then "Administrator guide"? Or the opposite?

Whatever you think is better, could you explain your thought process to make the decision?

Reviewer: Maybe we could clarify the proposal by mentioning some examples of existing pages that fit better in each section?

### 5. Community
  * #### Contributing
  * #### Testing
  * #### DCO
  * #### License
  * #### Code of Conduct
  * #### Release
  * #### Adopters
 

> # Alternatives

Reviewer: If we don't have alternatives, let's remove this heading :)

# Why

Restructuring the documentation will help in improving user experience and may save time for a user to search for relevant information effectively. This will encourage a newcomer to get familiar with the project in an efficient manner. The end goal of a good documentation is to follow the industrial best practices and provide accurate information to a user. 

I believe that by restructuring, we can group the related topics together in an organized way and use cross-references to link topics that are difficult to group but are relevant for understanding. This makes the documentation more user-friendly and also ensures uniform flow of content.

After we have a proper structure, it will become relatively easy to add information about new features. Maintainers will be able to save time on deciding the best place for adding new content. Also, improving documentation will help in improving Search Engine Optimization(SEO) and increase the engagement of the project.

# Pitfalls of the current solution

> A good documentation is one that is easy to understand for a newcomer and provides the exact amount of information that is needed according to the need. But looking at the current documentation structure, a lot of topics seem misplaced. For example, there is no need for a **"Contributing"** page in the prologue section. Prologue section should only give the introduction and the prerequisites for the project. Due to this, a user might need to put more effort to search for relevant information and decrease the user's productivity.

Reviewer 3: (nit) Correcting the sentence.
```suggestion
A good documentation is one that is easy to understand for a newcomer and provides the exact amount of information that is needed according to the need. But looking at the current documentation structure, a lot of topics seem misplaced. For example, there is no need for a **"Contributing"** page in the prologue section. Prologue section should only give the introduction and the prerequisites for the project. Due to this, a user might need to put more effort to search for relevant information and this might decrease the user's productivity.
```

# Goals

* To improve readability and make the documentation easier to understand.
* To ensure a uniform flow of content. 
* To remove irrelevant information from the documentation.
* To ensure that all new features get documented properly in the future.

# Audience

* Developers who rely on Prometheus Operator to facilitate the deployment, management and monitoring of Prometheus instances in their environment.
* Developers who want to set up alerts for their application.
* Administrators who want to monitor Prometheus and Alertmanager instances using Kubernetes tools.
* Maintainers who are responsible for optimizing Prometheus Operator.
* Contributors who want to engage and want to contribute effectively to the project.

# Non-Goals

* Adding extra documentation about new features, enhancements, troubleshooting, or other information that might be missing today. 

> # How

Reviewer 4: Do users always use the latest version of PO? If not, do we need different doc versions for different releases? Other open-source projects like Jaeger and Thanos do it:
https://www.jaegertracing.io/docs/1.57/
https://thanos.io/tip/thanos/

Author: For now we will mention the latest version but like you said, we need doc versions for each release. I think after we add missing concepts, we can also move for version specific documentation. Maybe, I should add this point in follow up. Wdyt ? @ArthurSens @nicolastakashi

Reviewer: I agree that versioned documentation would be awesome, I'm just a bit concerned with the scope of the proposal. If we decide to tackle everything at once, we might risk finishing the project on time.

Adding the idea as a follow-up sounds good!

I believe a basic structure for the website should be as given below:
### 1. Getting Started
  * **Introduction** - This section will introduce us to Prometheus-Operator and will talk about goals and the problems that are being solved by this project.
  * **Installation Guide** - This section will contain all the methods of installation of Prometheus-Operator in the production environment.
  * **Compatibility** - This section will provide information about compatibility of Prometheus-Operator with Kubernetes, Prometheus, Alertmanager and Thanos.
  * **Design** - This section will describe the design of Prometheus-Operator focusing on the various custom resource definitions(CRDs) it manages. 
### 2. API Reference
   This section will provide detailed information about different fields of the Custom Resources in Prometheus-Operator(config, spec, status and other information).    
### 3. User Guide

>    This section will provide focused information for developers who want to integrate Prometheus and Alertmanager using Prometheus-Operator in their applications along with examples and troubleshooting.

Author: Yeah thanks for mentioning it. If you find any more things like this to mention let me know here or in slack. I will note all these things down and mention it in the next proposal(after restructuring is properly done).

### 4. Administrator Guide
   This section will provide deeper insights for administrators who want to set up a robust, secure monitoring infrastructure. 
### 5. Community
  * **Contributing**
  * **Testing**
  * **DCO**
  * **License**
  * **Code of Conduct**
  * **Release**
  * **Adopters**

The reasoning behind the new structure is given below - 

> * The **“Getting-Started”** section should only serve information on how to begin your journey with the project. So, the first thing that comes to mind should be a basic overview followed by the installation. It won’t make sense to have the installation section appear in the User-Guide(in the current model) because installation is the foremost important thing to begin working on a project. With installation, users would also like to know the different releases and their compatibility, thus it makes more sense to add this topic just below the installation section. At last, Design doc should be present to give an overview of how different components interact and what their purpose is in a concise manner. 

Reviewer 4: > At last, Design doc should be present to give an overview of how different components interact and what their purpose is in a concise manner.

Agree. Besides text, I think drawing diagrams is also a good way to visualize the overview of how different components interact.

Author: Yeah, I have also seen an issue which I mentioned in my proposal where someone has made diagrams. Maybe we can incorporate those as well as make new ones.

> * API Reference is of extreme importance and is something needed by developers to implement the functionalities they need. It is something needed by both a **User** and an **Administrator**. As it is common for both personas, I think rather than duplicating this page in both sections, we should have a separate section for it due to its relevance. Further, looking at the current API reference page, it contains too much information for a single page. This makes the page look overly crowded and searching for any particular field becomes a tedious task. We can convert this page into multiple pages by grouping sections and cross-reference properly which will provide a better experience for a user.

Author: Yes, that is something we would need to check and then make changes accordingly.

* As mentioned in the audience, we also have people who want to contribute to the project. So, making a **“Community”** section serves the purpose of providing users to find how to contribute to the project. Also, it provides other community related information like release, adopters, and other topics which are not provided in the current model.

* We should remove the Kube-Prometheus section in the current model. We can mention the method of deploying Prometheus-Operator with Kube-Prometheus in the installation section due to its ease and popularity among users. But, as the helm-charts are not being maintained by Prometheus-Operator and have now been moved to Kube-Prometheus-Stack, it becomes difficult to manage and address the issues regarding Kube-Prometheus. 

# Action Plan

1. Making sections and organizing files as described above.
2. If needed, splitting the content of one file into multiple files to improve readability or changing the location of content wherever required.
3. Removing content that is unnecessary for the project.

# Follow-Ups

> After we successfully complete the Goals of this proposal, we can move on with the Non-Goals section. As mentioned, the next step would be to add the documentation for the topics that are currently missing.

Reviewer 4: One topic that's missing is ScrapeClass. Although it's a new experimental feature, I've seen a lot of users asking or even contributing to it. ScrapeClass solves some popular user needs, so I think we should soon have a document for it.

# How

I believe a basic structure for the website should be as given below:
### 1. Getting Started
  * #### Introduction
     This section will introduce us to Prometheus-Operator and will talk about goals of this project and the problems that are being solved by this project.
  * #### Installation Guide
     This section will contain all the methods of installation of Prometheus-Operator in you local environment.
### 2. API Reference
   This section will provide detailed information about different fields of the Custom Resources in Prometheus-Operator(config, spec, status and other information).    
### 3. User Guide
   This section will provide focused information for developers who want to integrate Prometheus and Alertmanager using Prometheus-Operator in their applications along with examples and trobleshooting.
### 4. Administrator Guide
   This section will provide deeper insights for administrators who want to setup a robust, secure monitoring infrastructure. 
### 5. Community
  * #### Contributing
  * #### Testing
  * #### DCO
  * #### License
  * #### Code of Conduct
  * #### Release

>   * #### Adopters

Author: Made a slight change from bulleted list. What do you think about this?

>  

Author: Made a slight change to bulleted list. What do you think about this?
